File: packages/socket/src/services/game/round-manager/round-recap.ts
// Per-round recap awards — extracted verbatim from RoundManager
// (round-manager.ts, Modul 1 of the SRP split). This was a private method
// with ZERO `this.` dependencies (a pure function of its arguments already),
// so no ctx is needed — only the `private` keyword and `this.` receiver moved.
import type {
  RoundRecapAward,
  RoundRecapKey,
} from "@razzoozle/common/types/game"

// Build up to 3 game-wide RoundRecapAward highlights for THIS round from the
// already-computed per-player intermediate rows (the aXxx fields + rankBefore
// + the freshly-unlocked achievements). Bots are excluded. Awards are picked
// by the SPEC priority list; we prefer variety (a clientId already used as a
// winner is de-prioritised but allowed when nothing else qualifies). Returns
// an empty array when nothing qualifies — the caller then omits the field so
// old clients are unaffected. NEVER throws.
export function computeRoundRecap(
  rows: ReadonlyArray<{
    clientId: string
    username: string
    avatar?: string
    isBot?: boolean
    aIsCorrect: boolean
    aResponseTimeMs: number | null
    aStreakAfter: number
    lastPoints: number
    answeredThisRound: boolean
  }>,
  rankAfterByClient: ReadonlyMap<string, number>,
  rankBefore: ReadonlyMap<string, number>,
  achievementsByClient: ReadonlyMap<string, string[]>,
  firstCorrectId: string | null,
  hasPriorRound: boolean,
): RoundRecapAward[] {
  try {
    // Only non-bot players are eligible for any highlight.
    const eligible = rows.filter((r) => !r.isBot)
    if (eligible.length === 0) {
      // ponytail: no human players this round — nothing to highlight.
      return []
    }

    const awards: RoundRecapAward[] = []
    const used = new Set<string>()

    const findRow = (clientId: string) =>
      eligible.find((r) => r.clientId === clientId)

    // Push one award if a winning row exists; prefer a not-yet-used winner but
    // accept an already-used one rather than skipping a real highlight. The
    // `pick` callback resolves the winning row (or null) given a candidate set.
    const add = (
      key: RoundRecapKey,
      candidates: ReadonlyArray<{ clientId: string; metric: number }>,
      order: "max" | "min",
      value: (metric: number) => number | undefined,
    ): void => {
      if (awards.length >= 3 || candidates.length === 0) {
        return
      }
      const better = (a: number, b: number) =>
        order === "max" ? a > b : a < b
      // Prefer a fresh winner; fall back to the best overall if all are used.
      const pickFrom = (
        pool: ReadonlyArray<{ clientId: string; metric: number }>,
      ): { clientId: string; metric: number } | null => {
        let best: { clientId: string; metric: number } | null = null
        for (const c of pool) {
          if (best === null || better(c.metric, best.metric)) {
            best = c
          }
        }
        return best
      }
      const fresh = candidates.filter((c) => !used.has(c.clientId))
      const winner = pickFrom(fresh) ?? pickFrom(candidates)
      if (winner === null) {
        return
      }
      const row = findRow(winner.clientId)
      if (row === undefined) {
        return
      }
      const v = value(winner.metric)
      awards.push({
        key,
        winnerName: row.username,
        ...(row.avatar !== undefined ? { winnerAvatar: row.avatar } : {}),
        ...(v !== undefined ? { value: v } : {}),
      })
      used.add(winner.clientId)
    }

    // 1. fastest_finger — correct answerer, smallest response time (ms).
    add(
      "fastest_finger",
      eligible
        .filter((r) => r.aIsCorrect && r.aResponseTimeMs !== null)
        .map((r) => ({
          clientId: r.clientId,
          metric: r.aResponseTimeMs!,
        })),
      "min",
      (ms) => ms,
    )

    // 2. first_correct — first correct answerer this round (arrival order). No value.
    if (
      awards.length < 3 &&
      firstCorrectId !== null &&
      findRow(firstCorrectId) !== undefined
    ) {
      add(
        "first_correct",
        [{ clientId: firstCorrectId, metric: 0 }],
        "max",
        () => undefined,
      )
    }

    // 3. streak — highest current streak >= 2 (value = streak length).
    add(
      "streak",
      eligible
        .filter((r) => r.aStreakAfter >= 2)
        .map((r) => ({ clientId: r.clientId, metric: r.aStreakAfter })),
      "max",
      (n) => n,
    )

    // 4. highest_round_score — most points gained THIS round (> 0).
    add(
      "highest_round_score",
      eligible
        .filter((r) => r.lastPoints > 0)
        .map((r) => ({ clientId: r.clientId, metric: r.lastPoints })),
      "max",
      (n) => n,
    )

    // 5. rank_climber — biggest positive rank improvement vs pre-round order.
    if (hasPriorRound) {
      const climbers: { clientId: string; metric: number }[] = []
      for (const r of eligible) {
        const before = rankBefore.get(r.clientId)
        const after = rankAfterByClient.get(r.clientId)
        if (before !== undefined && after !== undefined) {
          const climbed = before - after
          if (climbed > 0) {
            climbers.push({ clientId: r.clientId, metric: climbed })
          }
        }
      }
      add("rank_climber", climbers, "max", (n) => n)
    } else {
      // ponytail: round 1 has no pre-round order — rank_climber is meaningless.
    }

    // 6. achievement_unlock — a player who unlocked an achievement this round. No value.
    if (awards.length < 3) {
      const unlockers: { clientId: string; metric: number }[] = []
      for (const r of eligible) {
        const got = achievementsByClient.get(r.clientId)
        if (got !== undefined && got.length > 0) {
          unlockers.push({ clientId: r.clientId, metric: got.length })
        }
      }
      add("achievement_unlock", unlockers, "max", () => undefined)
    }

    // 7. slowest_player (filler) — largest response time among answerers (ms).
    add(
      "slowest_player",
      eligible
        .filter((r) => r.answeredThisRound && r.aResponseTimeMs !== null)
        .map((r) => ({
          clientId: r.clientId,
          metric: r.aResponseTimeMs!,
        })),
      "max",
      (ms) => ms,
    )

    // 8. most_wrong (filler) — per-round this is 0/1 wrong per player, so the
    // count is 1 for any wrong answerer; surfaced only as a last-resort filler.
    add(
      "most_wrong",
      eligible
        .filter((r) => r.answeredThisRound && !r.aIsCorrect)
        .map((r) => ({ clientId: r.clientId, metric: 1 })),
      "max",
      (n) => n,
    )

    return awards.slice(0, 3)
  } catch {
    // ponytail: recap is best-effort — any unexpected error yields no awards
    // and the caller omits the field (old clients keep working).
    return []
  }
}

File: rust/engine/src/round_recap.rs
use razzoozle_protocol::status::{RoundRecapAward, RoundRecapKey};
use std::collections::{HashMap, HashSet};

/// Per-round recap award row: minimal data needed to compute awards
#[derive(Debug, Clone)]
pub struct RoundRecapRow {
    pub client_id: String,
    pub username: String,
    pub avatar: Option<String>,
    pub is_bot: bool,
    pub correct: bool,
    pub response_time_ms: Option<i64>,
    pub streak_after: i32,
    pub last_points: i32,
    pub answered: bool,
}

/// Compute up to 3 game-wide award highlights from per-player data.
/// Bots are excluded; awards follow priority list.
/// Returns empty vec when nothing qualifies (old clients are unaffected).
pub fn compute_round_recap(
    rows: &[RoundRecapRow],
    rank_after_by_client: &HashMap<String, i32>,
    rank_before: &HashMap<String, i32>,
    first_correct_id: Option<&str>,
    has_prior_round: bool,
) -> Vec<RoundRecapAward> {
    let eligible: Vec<_> = rows.iter().filter(|r| !r.is_bot).collect();
    if eligible.is_empty() {
        return vec![];
    }

    let mut awards = vec![];
    let mut used = HashSet::new();

    // 1. fastest_finger — correct answerer, smallest response time
    {
        let mut candidates: Vec<_> = eligible
            .iter()
            .filter(|r| r.correct && r.response_time_ms.is_some())
            .map(|r| (r, r.response_time_ms.unwrap()))
            .collect();
        candidates.sort_by_key(|(_, ms)| *ms);

        if let Some((row, ms)) = candidates
            .iter()
            .find(|(r, _)| !used.contains(&r.username))
            .or_else(|| candidates.first())
        {
            awards.push(RoundRecapAward {
                key: RoundRecapKey::FastestFinger,
                winner_name: row.username.clone(),
                winner_avatar: row.avatar.clone(),
                value: Some(*ms as i32),
            });
            used.insert(row.username.clone());
        }
    }

    // 2. first_correct
    if awards.len() < 3 {
        if let Some(fid) = first_correct_id {
            if let Some(row) = eligible.iter().find(|r| r.client_id == fid) {
                if !used.contains(&row.username) {
                    awards.push(RoundRecapAward {
                        key: RoundRecapKey::FirstCorrect,
                        winner_name: row.username.clone(),
                        winner_avatar: row.avatar.clone(),
                        value: None,
                    });
                    used.insert(row.username.clone());
                }
            }
        }
    }

    // 3. streak — highest >= 2
    if awards.len() < 3 {
        let mut candidates: Vec<_> = eligible
            .iter()
            .filter(|r| r.streak_after >= 2)
            .map(|r| (r, r.streak_after as i64))
            .collect();
        candidates.sort_by(|a, b| b.1.cmp(&a.1)); // desc

        if let Some((row, streak)) = candidates
            .iter()
            .find(|(r, _)| !used.contains(&r.username))
            .or_else(|| candidates.first())
        {
            awards.push(RoundRecapAward {
                key: RoundRecapKey::Streak,
                winner_name: row.username.clone(),
                winner_avatar: row.avatar.clone(),
                value: Some(*streak as i32),
            });
            used.insert(row.username.clone());
        }
    }

    // 4. highest_round_score
    if awards.len() < 3 {
        let mut candidates: Vec<_> = eligible
            .iter()
            .filter(|r| r.last_points > 0)
            .map(|r| (r, r.last_points as i64))
            .collect();
        candidates.sort_by(|a, b| b.1.cmp(&a.1)); // desc

        if let Some((row, points)) = candidates
            .iter()
            .find(|(r, _)| !used.contains(&r.username))
            .or_else(|| candidates.first())
        {
            awards.push(RoundRecapAward {
                key: RoundRecapKey::HighestRoundScore,
                winner_name: row.username.clone(),
                winner_avatar: row.avatar.clone(),
                value: Some(*points as i32),
            });
            used.insert(row.username.clone());
        }
    }

    // 5. rank_climber
    if awards.len() < 3 && has_prior_round {
        let mut candidates: Vec<_> = eligible
            .iter()
            .filter_map(|r| {
                let before = rank_before.get(&r.client_id)?;
                let after = rank_after_by_client.get(&r.client_id)?;
                let climbed = before - after;
                if climbed > 0 {
                    Some((r, climbed as i64))
                } else {
                    None
                }
            })
            .collect();
        candidates.sort_by(|a, b| b.1.cmp(&a.1)); // desc

        if let Some((row, climbed)) = candidates
            .iter()
            .find(|(r, _)| !used.contains(&r.username))
            .or_else(|| candidates.first())
        {
            awards.push(RoundRecapAward {
                key: RoundRecapKey::RankClimber,
                winner_name: row.username.clone(),
                winner_avatar: row.avatar.clone(),
                value: Some(*climbed as i32),
            });
            used.insert(row.username.clone());
        }
    }

    // 6. slowest_player
    if awards.len() < 3 {
        let mut candidates: Vec<_> = eligible
            .iter()
            .filter(|r| r.answered && r.response_time_ms.is_some())
            .map(|r| (r, r.response_time_ms.unwrap()))
            .collect();
        candidates.sort_by(|a, b| b.1.cmp(&a.1)); // desc

        if let Some((row, ms)) = candidates
            .iter()
            .find(|(r, _)| !used.contains(&r.username))
            .or_else(|| candidates.first())
        {
            awards.push(RoundRecapAward {
                key: RoundRecapKey::SlowestPlayer,
                winner_name: row.username.clone(),
                winner_avatar: row.avatar.clone(),
                value: Some(*ms as i32),
            });
            used.insert(row.username.clone());
        }
    }

    // 7. most_wrong
    if awards.len() < 3 {
        if let Some(row) = eligible
            .iter()
            .find(|r| r.answered && !r.correct)
            .and_then(|first| {
                // Find unused, or just use first
                if !used.contains(&first.username) {
                    Some(*first)
                } else {
                    eligible
                        .iter()
                        .find(|r| r.answered && !r.correct)
                        .copied()
                }
            })
        {
            awards.push(RoundRecapAward {
                key: RoundRecapKey::MostWrong,
                winner_name: row.username.clone(),
                winner_avatar: row.avatar.clone(),
                value: Some(1),
            });
        }
    }

    awards
}
